File: csharp/sources/Valkey.Glide/Commands/Options/RestoreOptions.cs
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

using static Valkey.Glide.Commands.Constants.Constants;

namespace Valkey.Glide.Commands.Options;

/// <summary>
/// Optional arguments for the RESTORE command.
/// </summary>
/// <remarks>
/// IDLETIME and FREQ modifiers cannot be set at the same time.
/// See <a href="https://valkey.io/commands/restore/">valkey.io</a>
/// </remarks>
public class RestoreOptions
{
    /// <summary>
    /// When true, it represents REPLACE keyword has been used
    /// </summary>
    public bool HasReplace { get; set; } = false;

    /// <summary>
    /// It represents the idletime of object
    /// </summary>
    public long? Idletime { get; set; } = null;

    /// <summary>
    /// It represents the frequency of object
    /// </summary>
    public long? Frequency { get; set; } = null;

    /// <summary>
    /// Sets the REPLACE option
    /// </summary>
    /// <returns>This RestoreOptions instance for method chaining</returns>
    public RestoreOptions Replace()
    {
        HasReplace = true;
        return this;
    }

    /// <summary>
    /// Sets the IDLETIME option
    /// </summary>
    /// <param name="idletime">The idletime value</param>
    /// <returns>This RestoreOptions instance for method chaining</returns>
    public RestoreOptions SetIdletime(long idletime)
    {
        Idletime = idletime;
        return this;
    }

    /// <summary>
    /// Sets the FREQ option
    /// </summary>
    /// <param name="frequency">The frequency value</param>
    /// <returns>This RestoreOptions instance for method chaining</returns>
    public RestoreOptions SetFrequency(long frequency)
    {
        Frequency = frequency;
        return this;
    }

    /// <summary>
    /// Creates the argument array to be used in the RESTORE command
    /// </summary>
    /// <returns>A string array that holds the subcommands and their arguments</returns>
    /// <exception cref="ArgumentException">Thrown when both IDLETIME and FREQ are set</exception>
    internal GlideString[] ToArgs()
    {
        List<GlideString> resultList = [];

        if (HasReplace)
        {
            resultList.Add((GlideString)ReplaceKeyword);
        }

        if (Idletime.HasValue && Frequency.HasValue)
        {
            throw new ArgumentException("IDLETIME and FREQ cannot be set at the same time.");
        }

        if (Idletime.HasValue)
        {
            resultList.Add((GlideString)IdletimeKeyword);
            resultList.Add(Idletime.Value.ToGlideString());
        }

        if (Frequency.HasValue)
        {
            resultList.Add((GlideString)FreqKeyword);
            resultList.Add(Frequency.Value.ToGlideString());
        }

        return [.. resultList];
    }
}

File: java-old/client/src/main/java/glide/api/models/commands/RestoreOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands;

import static glide.api.models.GlideString.gs;

import glide.api.commands.GenericBaseCommands;
import glide.api.models.GlideString;
import java.util.ArrayList;
import java.util.List;
import lombok.*;

/**
 * Optional arguments to {@link GenericBaseCommands#restore(GlideString, long, byte[],
 * RestoreOptions)}.
 *
 * @see <a href="https://valkey.io/commands/restore/">valkey.io</a>
 * @apiNote <code>IDLETIME</code> and <code>FREQ</code> modifiers cannot be set at the same time.
 */
@Getter
@Builder
public final class RestoreOptions {
    /** <code>REPLACE</code> subcommand string to replace existing key */
    public static final String REPLACE_VALKEY_API = "REPLACE";

    /**
     * <code>ABSTTL</code> subcommand string to represent absolute timestamp (in milliseconds) for TTL
     */
    public static final String ABSTTL_VALKEY_API = "ABSTTL";

    /** <code>IDELTIME</code> subcommand string to set Object Idletime */
    public static final String IDLETIME_VALKEY_API = "IDLETIME";

    /** <code>FREQ</code> subcommand string to set Object Frequency */
    public static final String FREQ_VALKEY_API = "FREQ";

    /** When `true`, it represents <code>REPLACE</code> keyword has been used */
    @Builder.Default private boolean hasReplace = false;

    /** When `true`, it represents <code>ABSTTL</code> keyword has been used */
    @Builder.Default private boolean hasAbsttl = false;

    /** It represents the idletime of object */
    @Builder.Default private Long idletime = null;

    /** It represents the frequency of object */
    @Builder.Default private Long frequency = null;

    /**
     * Creates the argument to be used in {@link GenericBaseCommands#restore(GlideString, long,
     * byte[], RestoreOptions)}
     *
     * @return a <code>GlideString</code> array that holds the subcommands and their arguments.
     */
    public GlideString[] toArgs() {
        List<GlideString> resultList = new ArrayList<>();

        if (hasReplace) {
            resultList.add(gs(REPLACE_VALKEY_API));
        }

        if (hasAbsttl) {
            resultList.add(gs(ABSTTL_VALKEY_API));
        }

        if (idletime != null && frequency != null) {
            throw new IllegalArgumentException("IDLETIME and FREQ cannot be set at the same time.");
        }

        if (idletime != null) {
            resultList.add(gs(IDLETIME_VALKEY_API));
            resultList.add(gs(Long.toString(idletime)));
        }

        if (frequency != null) {
            resultList.add(gs(FREQ_VALKEY_API));
            resultList.add(gs(Long.toString(frequency)));
        }

        return resultList.toArray(new GlideString[0]);
    }

    /** Custom setter methods for replace and absttl */
    public static class RestoreOptionsBuilder {
        public RestoreOptionsBuilder replace() {
            return hasReplace(true);
        }

        public RestoreOptionsBuilder absttl() {
            return hasAbsttl(true);
        }
    }
}
